Print AIR in WASM BBQ disassembly

justinmichaud opened the pull request:
#### ad2e13b2a8e69c2ea538cdb8c7cb101a40555266
<pre>
Print AIR in WASM BBQ disassembly
<a href="https://bugs.webkit.org/show_bug.cgi?id=244644">https://bugs.webkit.org/show_bug.cgi?id=244644</a>

Reviewed by Yusuke Suzuki.

1) Print AIR in WASM disassembly

2) Add extra jit comments, notably the compiler backtrace where each patchpoint is constructed

3) Add support for multiple jit comments

4) Add support for printing jit comments on x86

* Source/JavaScriptCore/assembler/LinkBuffer.cpp:
(JSC::LinkBuffer::linkComments):
* Source/JavaScriptCore/b3/B3PatchpointSpecial.cpp:
(JSC::B3::PatchpointSpecial::generate):
* Source/JavaScriptCore/b3/B3Value.h:
* Source/JavaScriptCore/disassembler/X86Disassembler.cpp:
(JSC::tryToDisassemble):
* Source/JavaScriptCore/jit/AssemblyHelpers.cpp:
(JSC::AssemblyHelpers::restoreCalleeSavesFromEntryFrameCalleeSavesBuffer):
(JSC::AssemblyHelpers::restoreCalleeSavesFromVMEntryFrameCalleeSavesBufferImpl):
(JSC::AssemblyHelpers::loadWasmContextInstance):
(JSC::AssemblyHelpers::storeWasmContextInstance):
(JSC::AssemblyHelpers::debugCall):
(JSC::AssemblyHelpers::copyCalleeSavesToEntryFrameCalleeSavesBufferImpl):
(JSC::AssemblyHelpers::emitSave):
(JSC::AssemblyHelpers::emitRestore):
(JSC::AssemblyHelpers::emitSaveCalleeSavesFor):
(JSC::AssemblyHelpers::emitRestoreCalleeSavesFor):
* Source/JavaScriptCore/jit/AssemblyHelpersSpoolers.h:
(JSC::AssemblyHelpers::Spooler::execute):
* Source/JavaScriptCore/jit/ScratchRegisterAllocator.cpp:
(JSC::ScratchRegisterAllocator::preserveReusedRegistersByPushing):
(JSC::ScratchRegisterAllocator::restoreReusedRegistersByPopping):
(JSC::ScratchRegisterAllocator::preserveRegistersToStackForCall):
(JSC::ScratchRegisterAllocator::restoreRegistersFromStackForCall):
* Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp:
(JSC::Wasm::AirIRGenerator::addCall):
(JSC::Wasm::parseAndCompileAir):
* Source/JavaScriptCore/wasm/WasmBBQPlan.cpp:
(JSC::Wasm::BBQPlan::didCompleteCompilation):
* Source/JavaScriptCore/wasm/WasmThunks.cpp:
(JSC::Wasm::throwExceptionFromWasmThunkGenerator):
(JSC::Wasm::throwStackOverflowFromWasmThunkGenerator):
(JSC::Wasm::triggerOMGEntryTierUpThunkGenerator):
* Source/JavaScriptCore/wasm/js/JSToWasm.cpp:
(JSC::Wasm::marshallJSResult):
(JSC::Wasm::createJSToWasmWrapper):
* Source/WTF/wtf/Assertions.h:

Canonical link: <a href="https://commits.webkit.org/254198@main">https://commits.webkit.org/254198@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/185e739ade9865ad257a164dd67edbc6d98e6d80

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/88070 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/32235 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/18784 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/97213 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/152707 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/30601 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/26541 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/80186 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/91952 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/93680 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/24664 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/74714 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/24645 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/79586 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/79707 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/67630 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/79821 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/28230 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/13591 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/73574 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/28329 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/14582 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/26199 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2955 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/31356 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/37470 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/76418 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/30304 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/33818 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/16960 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

justinmichaud commented on Source/JavaScriptCore/wasm/WasmBBQPlan.cpp:
```diff
@@ -265,8 +265,22 @@ void BBQPlan::didCompleteCompilation()
+
             function->entrypoint.compilation = makeUnique<Compilation>(
-                FINALIZE_CODE(linkBuffer, JITCompilationPtrTag, "WebAssembly BBQ function[%i] %s name %s", functionIndex, signature.toString().ascii().data(), makeString(IndexOrName(functionIndexSpace, m_moduleInformation->nameSection->get(functionIndexSpace))).ascii().data()),
+                FINALIZE_CODE_IF(false, linkBuffer, JITCompilationPtrTag, "WebAssembly BBQ function[%i] %s name %s", functionIndex, signature.toString().ascii().data(), makeString(IndexOrName(functionIndexSpace, m_moduleInformation->nameSection->get(functionIndexSpace))).ascii().data()),
```
Is there a better way?

Constellation replied:
I think it is OK to just use the `FINALIZE_CODE`, as what we are doing in FTL is the same.

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQPlan.cpp:
```diff
@@ -265,8 +265,22 @@ void BBQPlan::didCompleteCompilation()
 
             computePCToCodeOriginMap(context, linkBuffer);
 
+            if (shouldDumpDisassemblyFor(CompilationMode::BBQMode)) {
```
Let's put `UNLIKELY`.

saambarati commented on Source/JavaScriptCore/b3/B3PatchpointSpecial.cpp:
```diff
@@ -162,13 +164,16 @@ MacroAssembler::Jump PatchpointSpecial::generate(Inst& inst, CCallHelpers& jit,
 
     StackmapGenerationParams params(value, reps, context);
+    jit.comment("Patchpoint body start, with arg value reps: ", listDump(reps));
+    jit.comment("generated by: ", value->sourceString());
```
We shouldn’t be evaluating things as arguments to this method

saambarati replied:
I guess it’s just a getter. But it feels like bad form

Constellation replied:
Probably we should introduce `jitComment(jit, ...)` macro, which does not evaluate the arguments when comment is disabled.

justinmichaud replied:
Comment is always inline, and does not consume its arguments when assertions are disabled

justinmichaud replied:
Ooops, I am wrong

saambarati commented on Source/JavaScriptCore/b3/B3Value.h:
```diff
@@ -730,7 +730,34 @@ class JS_EXPORT_PRIVATE Value {
     NO_RETURN_DUE_TO_CRASH static void badKind(Kind, unsigned);
 
+#if ASSERT_ENABLED
+    String m_cppSourceString = generateSourceString();
```
Let’s use initializer syntax

saambarati commented on Source/JavaScriptCore/wasm/WasmBBQPlan.cpp:
```diff
@@ -265,8 +265,22 @@ void BBQPlan::didCompleteCompilation()
+            if (shouldDumpDisassemblyFor(CompilationMode::BBQMode)) {
+                auto* disassembler = context.procedure->code().disassembler();
+
+                const char* airPrefix = "Air" "    ";
```
Why as two strings concatenated? Not sure how this is easier to read

saambarati commented on Source/JavaScriptCore/b3/B3Value.h:
```diff
@@ -730,7 +730,34 @@ class JS_EXPORT_PRIVATE Value {
+#if ASSERT_ENABLED
+    String m_cppSourceString = generateSourceString();
+
+    static String generateSourceString()
```
Is this really printing a source string? Or just a frame name?

Constellation commented on Source/JavaScriptCore/wasm/generateWasmOpsHeader.py:
```diff
@@ -253,6 +253,14 @@ def atomicMemoryLog2AlignmentGenerator(filter):
         return static_cast<bool>(nullable);
     }
 
+    void dump(PrintStream& out) const {
```
Insert new-line before `{`.

Constellation commented on Source/JavaScriptCore/assembler/AbstractMacroAssembler.h:
```diff
@@ -57,6 +57,8 @@ namespace DFG {
 struct OSRExit;
 }
 
+#define JIT_COMMENT(...) if (Options::dumpDisassembly()) { jit.comment(__VA_ARGS__); } else { (void) jit; }
```
Let's take `jit` as a first parameter. Some place uses `m_jit`, some place uses `JIT`'s pointer.

Constellation replied:
Also, let's use `UNLIKELY` for `Options::dumpDisassembly()`.
And let's wrap everything with `do { } while (0)` as a normal C macro's convention to prevent accidental `else` etc chaining.
```
#define JIT_COMMENT(jit, ...) \
    do { \
        if (UNLIKELY(Options::dumpDisassembly())) \
            (jit).comment(__VA_ARGS__); \
    } while (0)
```

Constellation commented on Source/JavaScriptCore/jit/AssemblyHelpers.cpp:
```diff
@@ -1098,6 +1104,7 @@ bool AssemblyHelpers::storeWasmContextInstanceNeedsMacroScratchRegister()
 
 void AssemblyHelpers::debugCall(VM& vm, V_DebugOperation_EPP function, void* argument)
 {
+    comment("debugCall");
```
Let's change it to use JIT_COMMENT. Passing *this.

Constellation commented on Source/JavaScriptCore/jit/AssemblyHelpers.cpp:
```diff
@@ -1142,6 +1149,7 @@ void AssemblyHelpers::debugCall(VM& vm, V_DebugOperation_EPP function, void* arg
 
 void AssemblyHelpers::copyCalleeSavesToEntryFrameCalleeSavesBufferImpl(GPRReg calleeSavesBuffer)
 {
+    comment("copyCalleeSavesToEntryFrameCalleeSavesBufferImpl ", calleeSavesBuffer);
```
Ditto.

Constellation commented on Source/JavaScriptCore/jit/AssemblyHelpers.cpp:
```diff
@@ -1255,6 +1263,7 @@ void AssemblyHelpers::cageConditionallyAndUntag(Gigacage::Kind kind, GPRReg stor
 
 void AssemblyHelpers::emitSave(const RegisterAtOffsetList& list)
 {
+    comment("emitSave ", list);
```
Ditto.

Constellation commented on Source/JavaScriptCore/jit/AssemblyHelpers.cpp:
```diff
@@ -1274,6 +1283,7 @@ void AssemblyHelpers::emitSave(const RegisterAtOffsetList& list)
 
 void AssemblyHelpers::emitRestore(const RegisterAtOffsetList& list)
 {
+    comment("emitRestore ", list);
```
Ditto.

Constellation commented on Source/JavaScriptCore/jit/AssemblyHelpers.cpp:
```diff
@@ -1295,6 +1305,7 @@ void AssemblyHelpers::emitSaveCalleeSavesFor(const RegisterAtOffsetList* calleeS
 {
     RegisterSet dontSaveRegisters = RegisterSet(RegisterSet::stackRegisters());
     unsigned registerCount = calleeSaves->registerCount();
+    comment("emitSaveCalleeSavesFor ", *calleeSaves, " dontRestore: ", dontSaveRegisters);
```
Ditto.

Constellation commented on Source/JavaScriptCore/jit/AssemblyHelpers.cpp:
```diff
@@ -1321,6 +1332,7 @@ void AssemblyHelpers::emitRestoreCalleeSavesFor(const RegisterAtOffsetList* call
 {
     RegisterSet dontRestoreRegisters = RegisterSet(RegisterSet::stackRegisters());
     unsigned registerCount = calleeSaves->registerCount();
+    comment("emitRestoreCalleeSavesFor ", *calleeSaves, " dontSave: ", dontRestoreRegisters);
```
Ditto.

Constellation commented on Source/JavaScriptCore/jit/AssemblyHelpers.cpp:
```diff
@@ -1056,24 +1058,28 @@ AssemblyHelpers::JumpList AssemblyHelpers::branchIfValue(VM& vm, JSValueRegs val
     }
 #endif
     move(src, Wasm::PinnedRegisterInfo::get().wasmContextInstancePointer);
+    comment("Store wasm context instance done");
```
Ditto.

Constellation commented on Source/JavaScriptCore/jit/AssemblyHelpers.cpp:
```diff
@@ -1056,24 +1058,28 @@ AssemblyHelpers::JumpList AssemblyHelpers::branchIfValue(VM& vm, JSValueRegs val
 
 void AssemblyHelpers::storeWasmContextInstance(GPRReg src)
 {
+    comment("Store wasm context instance from", src);
```
Ditto.

Constellation commented on Source/JavaScriptCore/jit/AssemblyHelpers.cpp:
```diff
@@ -1056,24 +1058,28 @@ AssemblyHelpers::JumpList AssemblyHelpers::branchIfValue(VM& vm, JSValueRegs val
     }
 #endif
     move(Wasm::PinnedRegisterInfo::get().wasmContextInstancePointer, dst);
+    comment("Load wasm instance done");
```
Ditto.

Constellation commented on Source/JavaScriptCore/jit/AssemblyHelpers.cpp:
```diff
@@ -1056,24 +1058,28 @@ AssemblyHelpers::JumpList AssemblyHelpers::branchIfValue(VM& vm, JSValueRegs val
 #if ENABLE(WEBASSEMBLY)
 void AssemblyHelpers::loadWasmContextInstance(GPRReg dst)
 {
+    comment("Load wasm context instance to ", dst);
```
Ditto.

Constellation commented on Source/JavaScriptCore/jit/AssemblyHelpers.cpp:
```diff
@@ -758,6 +759,7 @@ void AssemblyHelpers::restoreCalleeSavesFromVMEntryFrameCalleeSavesBufferImpl(GP
     addPtr(TrustedImm32(EntryFrame::calleeSaveRegistersBufferOffset()), entryFrameGPR);
 
     RegisterAtOffsetList* allCalleeSaves = RegisterSet::vmCalleeSaveRegisterOffsets();
+    comment("restoreCalleeSavesFromVMEntryFrameCalleeSavesBufferImpl ", entryFrameGPR, " callee saves: ", *allCalleeSaves, " skip: ", skipList);
```
Ditto.

Constellation commented on Source/JavaScriptCore/jit/AssemblyHelpers.cpp:
```diff
@@ -672,6 +672,7 @@ void AssemblyHelpers::restoreCalleeSavesFromEntryFrameCalleeSavesBuffer(EntryFra
     RegisterAtOffsetList* allCalleeSaves = RegisterSet::vmCalleeSaveRegisterOffsets();
     RegisterSet dontRestoreRegisters = RegisterSet::stackRegisters();
     unsigned registerCount = allCalleeSaves->registerCount();
+    comment("restoreCalleeSavesFromEntryFrameCalleeSavesBuffer ", *allCalleeSaves, " skip: ", dontRestoreRegisters);
```
Ditto.

Constellation reviewed (approved):
r=me with comments.

Commit:
Print AIR in WASM BBQ disassembly
https://bugs.webkit.org/show_bug.cgi?id=244644

Reviewed by Yusuke Suzuki.

1) Print AIR in WASM disassembly

2) Add extra jit comments, notably the compiler backtrace where each patchpoint is constructed

3) Add support for multiple jit comments

4) Add support for printing jit comments on x86

* Source/JavaScriptCore/assembler/LinkBuffer.cpp:
(JSC::LinkBuffer::linkComments):
* Source/JavaScriptCore/b3/B3PatchpointSpecial.cpp:
(JSC::B3::PatchpointSpecial::generate):
* Source/JavaScriptCore/b3/B3Value.h:
* Source/JavaScriptCore/disassembler/X86Disassembler.cpp:
(JSC::tryToDisassemble):
* Source/JavaScriptCore/jit/AssemblyHelpers.cpp:
(JSC::AssemblyHelpers::restoreCalleeSavesFromEntryFrameCalleeSavesBuffer):
(JSC::AssemblyHelpers::restoreCalleeSavesFromVMEntryFrameCalleeSavesBufferImpl):
(JSC::AssemblyHelpers::loadWasmContextInstance):
(JSC::AssemblyHelpers::storeWasmContextInstance):
(JSC::AssemblyHelpers::debugCall):
(JSC::AssemblyHelpers::copyCalleeSavesToEntryFrameCalleeSavesBufferImpl):
(JSC::AssemblyHelpers::emitSave):
(JSC::AssemblyHelpers::emitRestore):
(JSC::AssemblyHelpers::emitSaveCalleeSavesFor):
(JSC::AssemblyHelpers::emitRestoreCalleeSavesFor):
* Source/JavaScriptCore/jit/AssemblyHelpersSpoolers.h:
(JSC::AssemblyHelpers::Spooler::execute):
* Source/JavaScriptCore/jit/ScratchRegisterAllocator.cpp:
(JSC::ScratchRegisterAllocator::preserveReusedRegistersByPushing):
(JSC::ScratchRegisterAllocator::restoreReusedRegistersByPopping):
(JSC::ScratchRegisterAllocator::preserveRegistersToStackForCall):
(JSC::ScratchRegisterAllocator::restoreRegistersFromStackForCall):
* Source/JavaScriptCore/wasm/WasmAirIRGenerator.cpp:
(JSC::Wasm::AirIRGenerator::addCall):
(JSC::Wasm::parseAndCompileAir):
* Source/JavaScriptCore/wasm/WasmBBQPlan.cpp:
(JSC::Wasm::BBQPlan::didCompleteCompilation):
* Source/JavaScriptCore/wasm/WasmThunks.cpp:
(JSC::Wasm::throwExceptionFromWasmThunkGenerator):
(JSC::Wasm::throwStackOverflowFromWasmThunkGenerator):
(JSC::Wasm::triggerOMGEntryTierUpThunkGenerator):
* Source/JavaScriptCore/wasm/js/JSToWasm.cpp:
(JSC::Wasm::marshallJSResult):
(JSC::Wasm::createJSToWasmWrapper):
* Source/WTF/wtf/Assertions.h:

Canonical link: https://commits.webkit.org/254198@main
